packages/trace-viewer/src/ui/aiConversation.tsx

```diff
@@ -17,39 +17,66 @@ import { useCallback, useState } from 'react';
 import Markdown from 'markdown-to-jsx';
 import './aiConversation.css';
 import { clsx } from '@web/uiUtils';
-import { useLLMConversation } from './llm';
+import { LLMMessage, useLLMChat, useLLMConversation } from './llm';
 
-export function AIConversation({ conversationId }: { conversationId: string }) {
+export function AIConversation({ conversationId, setConversationId }: { conversationId?: string, setConversationId(id: string): void; }) {
+  const chat = useLLMChat();
   const [history, conversation] = useLLMConversation(conversationId);
-  const [input, setInput] = useState('');
 
+  if (!conversation) {
+    return <AIConversationView
+      history={history}
+      apiName={chat.api.name}
+      onSend={content => {
+        const id = crypto.randomUUID();
+        const conversation = chat.startConversation(id, '');
+        conversation.send(content);
+        setConversationId(id);
+      }}
+    />;
+  }
+
+  return (
+    <AIConversationView
+      history={history}
+      apiName={chat.api.name}
+      onSend={content => conversation.send(content)}
+      onCancel={conversation.isSending() ? () => conversation.abortSending() : undefined}
+    />
+  );
+}
+
+export function AIConversationView({ history, onSend, onCancel, apiName }: { history: LLMMessage[], onSend(message: string): void, onCancel?(): void; apiName: string; }) {
+  const [input, setInput] = useState('');
   const onSubmit = useCallback(() => {
     setInput(content => {
-      conversation.send(content);
+      onSend(content);
       return '';
     });
-  }, [conversation]);
+  }, [onSend]);
 
   return (
-    <div className='chat-container'>
-      <p className='chat-disclaimer'>Chat based on {conversation.chat.api.name}. Check for mistakes.</p>
-      <hr/>
-      <div className='messages-container'>
-        {history.filter(({ role }) => role !== 'developer').map((message, index) => (
-          <div
-            key={'' + index}
-            className={clsx('message', message.role === 'user' && 'user-message')}
-          >
-            {message.role === 'assistant' && (
-              <div className='message-icon'>
-                <img src='playwright-logo.svg' />
+    <div className='chat-container' role='region' aria-label='Chat'>
+      <div className='chat-scroll'>
+        <p className='chat-disclaimer'>Chat based on {apiName}. Check for mistakes.<hr/></p>
+
+        <div className='messages-container'>
+          {history.filter(({ role }) => role !== 'developer').map((message, index) => (
+            <div
+              key={'' + index}
```

Comment:
This should be using the message ID returned from the API. I don't know why you wrote this, but it looks like you're trying to get around some rule or something. In almost no scenario should you use `index` as `key`.